src/ucp/tag/eager_rcv.c

```diff
@@ -74,22 +74,24 @@ ucp_eager_handler(void *arg, void *data, size_t length, void *desc,
         }
     }
 
+    rdesc = ucp_recv_desc_get(worker, data, desc, length, hdr_len, flags);
+    if (!rdesc) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
     ucs_trace_req("unexp recv %c%c%c tag %"PRIx64" length %zu desc %p",
                   (flags & UCP_RECV_DESC_FLAG_FIRST) ? 'f' : '-',
                   (flags & UCP_RECV_DESC_FLAG_LAST)  ? 'l' : '-',
                   (flags & UCP_RECV_DESC_FLAG_EAGER) ? 'e' : '-',
                   recv_tag, length, rdesc);
 
-    if (data != rdesc + 1) {
-        memcpy(rdesc + 1, data, length);
-    }
-
-    rdesc->length  = length;
-    rdesc->hdr_len = hdr_len;
-    rdesc->flags   = flags;
     ucs_queue_push(&context->tag.unexpected, &rdesc->queue);
 
-    status = UCS_INPROGRESS;
+    if (flags & UCP_RECV_DESC_FLAG_SYNC) {
```

Comment:
In case of expected match `ucp_eager_handler` returns OK then upper level CBs (`ucp_eager_sync_first_handler` and `ucp_eager_sync_only_handler`) send ACK, but in case of unexpected it's early do ACK before the data is not in user's buffer. As I mentioned above, when I remove copying of descriptors on UCT level (by handling `UCS_INPROGRESS` and `UCT_AM_FLAG_DESC`) any UCP CB must return `UCS_OK` if it's not supported by a transport